compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp

```diff
@@ -96,6 +97,79 @@ static bool getTilingOptionsFromConfig(func::FuncOp funcOp, int64_t tilingLevel,
   return false;
 }
 
+/// Computes the canonical shape used to vectorize this dispatch. Retrieves
+/// the vectorization tile sizes (parallel and reduction levels) out of the
+/// lowering config and adjusts them to the format expected by the Linalg
+/// vectorizer.
+static SmallVector<int64_t> getCanonicalVectorShape(func::FuncOp funcOp) {
+  FailureOr<Operation *> rootOp = getRootOp(funcOp);
+  if (failed(rootOp)) {
+    return {};
+  }
+
+  unsigned numTileLevels =
+      mlir::iree_compiler::getNumTileLevels(rootOp.value());
+  if (numTileLevels < 3) {
+    return {};
+  }
+
+  // Retrieve the tile sizes from the last two tiling levels (parallel and
+  // reduction) used for vectorization.
+  SmallVector<int64_t> canonicalVectorShape =
+      mlir::iree_compiler::getTileSizes(rootOp.value(), numTileLevels - 2);
+  SmallVector<int64_t> reductionTileSizes =
+      mlir::iree_compiler::getTileSizes(rootOp.value(), numTileLevels - 1);
+
+  if (!reductionTileSizes.empty()) {
+    assert(canonicalVectorShape.size() == reductionTileSizes.size() &&
+           "Unexpected tile sizes");
+
+    // Combine the reduction tile sizes with the parallel tile sizes already in
+    // the canonical vector shape.
+    for (int i = 0, end = canonicalVectorShape.size(); i < end; ++i) {
+      if (reductionTileSizes[i] > 0)
+        canonicalVectorShape[i] = reductionTileSizes[i];
+    }
+  }
+
+  // Replace zeros in canonical vector shape to turn it into a valid shape.
+  std::replace(canonicalVectorShape.begin(), canonicalVectorShape.end(), 0, 1);
+  return canonicalVectorShape;
+}
+
+// Give the canonical vector shape of a dispatch, returns the vector sizes for a
+// particular linalg op within that dispatch.
+static SmallVector<int64_t> getVectorSizes(
+    linalg::LinalgOp linalgOp, ArrayRef<int64_t> canonicalVectorShape) {
+  FailureOr<Operation *> rootOp =
+      getRootOp(linalgOp->getParentOfType<func::FuncOp>());
+  if (failed(rootOp)) {
+    return {};
+  }
+
+  // TODO: Not easy to infer the tiles sizes for an op that is not the root op.
+  if (*rootOp != linalgOp.getOperation()) {
+    return {};
+  }
+
+  // TODO: Masking is only supported for dynamic shapes.
```

Comment:
TODO is more like an action item. We can comment about what is going to be done, not describing what's happening here. I'd suggest to update it to `TODO: Support masking for non dynamic shapes.`